public/index.jsx

```diff
@@ -0,0 +1,65 @@
+/* eslint-disable react/prop-types, react/no-multi-comp, react/jsx-no-bind */
+import React from 'react';
+import { render } from 'react-dom';
+import { BrowserRouter, Match, Miss, Redirect } from 'react-router';
+
+import './style.scss';
+import Header from './components/header';
+import Login from './components/login';
+import Navbar from './components/navbar';
+import NotFound from './components/not-found';
+import Profile from './components/profile';
+import { auth } from './utils';
+
+const renderBasedOnAuth = (Component, pattern, props) => {
+  if (auth.isAuthenticated() && auth.userCanAccess(pattern)) {
+    return <Component />;
+  } else if (auth.isAuthenticated()) {
+    return <NotFound />;
+  } else {
+    return (
+      <Redirect to={{
+        pathname: '/login',
+        stats: { from: props.location }
+      }} />
+    );
+  }
+};
+
+const MatchWhenNotLoggedIn = ({ component: Component, pattern, ...rest }) => (
+  <Match {...rest} pattern={pattern} render={() => {
+    if (!auth.isAuthenticated()) return <Component />;
+    return <Redirect to={{ pathname: '/' }} />;
+  }} />
+);
+
+const MatchWhenAuthorized = ({ component: Component, pattern, ...rest }) => (
+  <Match {...rest} pattern={pattern} render={props => (renderBasedOnAuth(Component, pattern, props))} />
+);
+
+const handleLogout = (router) => {
+  auth.logout();
+  router.transitionTo('/');
+};
+
+const App = () => {
+  return (
+    <BrowserRouter>
+      {
+        ({ router }) => (
+          <div className="App">
+            <Header />
+            <Navbar onLogout={handleLogout.bind(null, router)} />
```

Comment:
Components that need access to the user, should be given the user by this component. Examples right now would be `Navbar` and `Profile`